Lazy instantiation and proxy for issue #143

overclokk opened the pull request:
This is the implementation for the ProxyManager I made for using with Auryn\Injector.

I updated the readme, I'm not native english and I may have made some grammatical errors.

I preferred to add new class with an interface for wrapping the ProxyManager instead of injecting directly the class.

The new implementation is tested, I created a dedicated file for it.

Updated the composer.json to support PHP 7.2 >= and ProxyManager ~2.2.* (that is the last version for the 7.2)

Updated travis.yml to test with PHP 7.2|7.3|7.4|nightly and added some tweak

In future there are also PHPUnit that needs to be updated to the latest version and change athletic/athletic with phpbench/phpbench because athletic is no longer supported, for now they are still working fine.

kelunik commented on lib/Injector.php:
```diff
@@ -15,42 +16,46 @@ class Injector
+    const M_DELEGATE_ARGUMENT = '%s::delegate expects a valid callable or executable class::method string at Argument 2%s';
     const E_CYCLIC_DEPENDENCY = 11;
-    const M_CYCLIC_DEPENDENCY = "Detected a cyclic dependency while provisioning %s";
+    const M_CYCLIC_DEPENDENCY = 'Detected a cyclic dependency while provisioning %s';
```
Please revert the code style changes for easier review.

overclokk replied:
They was fixed by the php-cs-fixer, I thought it was correct to run it before sending a pr.

kelunik replied:
I'm currently just on my phone, but I'll have to look which config this repository currently uses. We can fix the code style in another PR or directly on master then.

overclokk replied:
How do you suggest to proceed?

overclokk replied:
I think I did it

kelunik commented on lib/Injector.php:
```diff
@@ -15,42 +16,46 @@ class Injector
     const M_MAKING_FAILED = "Making %s did not result in an object, instead result is of type '%s'";
 
     private $reflector;
+    private $proxy_manager;
```
Please adhere to camel Case as all other properties.

overclokk replied:
Ok

kelunik commented on lib/Injector.php:
```diff
@@ -369,6 +404,12 @@ public function make($name, array $args = array())
+                    $this->prepares_proxy[$normalizedClass] = $this->prepares[$normalizedClass];
+                }
+                $obj = $this->resolveProxy($className, $normalizedClass, $args);
+                unset($this->prepares[$normalizedClass]);
```
Why are those unset?

overclokk replied:
Because preparation is made inside the `resolveProxy()' with 'prepareInstance()' when the instantiation of the real object is called at runtime, this happens after the 'make()' method is called, those unset make sure that the 'prepareInstance()' called on the line 417 will be bypassed without any error.

kelunik commented:
@overclokk: I've thought a bit about possible implementations and have implemented my idea in #185, what do you think? It's a generic proxy approach there without a dependency on `proxy-manager`. I'm not sure whether the proxy callable should receive a callable acting as constructor or the actual constructor arguments.

overclokk commented:
@kelunik 

> I've thought a bit about possible implementations and have implemented my idea in #185, what do you think? It's a generic proxy approach there without a dependency on `proxy-manager`.

I like that idea, maybe I can add more generic approach on my implementation following yours, I have to think about it and do some test.

In your implementation the `::prepare()`  method call the real object as soon as you call `$injector->prepare()`

Here part of the code to replicate the issue
```php

class V8 implements Engine
{
	public $value = '';

    public function __construct()
    {
        print __METHOD__ . PHP_EOL;
    }

    public function turnOn()
    {
        print __METHOD__ . PHP_EOL;
    }
}

$injector = new Auryn\Injector;
$injector->alias(Engine::class, V8::class);
$injector->proxy(Car::class, $proxy(Car::class));
$injector->proxy(V8::class, $proxy(V8::class));
$injector->prepare( V8::class, function ( V8 $v8, \Auryn\Injector $injector ) {
	$v8->value = 42;
} );
```

overclokk commented:
@kelunik I implemented yours idea from #185 , I just simplified the closure, all tests passes and no more dependency from ProxyManager, what do you think?

overclokk commented:
@kelunik have you seen the code I made?

kelunik commented:
@overclokk: Not yet, was on holiday, but will probably have time in the coming days.

overclokk commented:
@kelunik Ok, no problem :-)

overclokk commented:
@kelunik maybe after one year you are no more interested in this PR, if this is the case feel free to close it and I will continue to maintain the forked version, don't worry, if you are still interested let me know.

kelunik commented:
@overclokk Thanks for the reminder. I'm still interested in supporting the feature and have now started a larger cleanup in https://github.com/amphp/injector.

overclokk commented:
@kelunik I'm happy to hear that.
So now all PR should ba made to the new repository?

kelunik commented:
Yes, will still accept new PRs for bug fixes, but new features will land in the new repo. Don't worry, I'll merge the PR manually, you don't need to do anything.

overclokk commented:
Ok, perfect :-)
Thank you

kelunik commented:
@overclokk Merged! :rocket:

overclokk commented:
This is awesome, thank you :-)
